File: translator.py
from PyQt5.QtCore import QThread, pyqtSignal
from cache import Cache
import openai

class Translator(QThread):
    signal = pyqtSignal(str)

    def __init__(self, text:str, api_key=None, base_url=None, model="gpt-4.1-nano", prompt="Translate the following text to Chinese:", cache:Cache=None):
        super().__init__()

        self.text = text
        self.api_key = api_key
        self.base_url = base_url
        self.model = model
        self.prompt = prompt
        self.cache = cache
      
    def run(self):
        try:
            # 检查是否被请求中断
            if self.isInterruptionRequested():
                return
                
            # 检查缓存
            cache_text = self.cache.get(self.text, gap=3)
            if cache_text:
                self.signal.emit(cache_text)
                return
                
            # 如果没有API密钥，提示错误
            if not self.api_key:
                self.signal.emit('@An error occurred:API密钥未设置，请在设置中配置API。\n ')
                return
                
            # 如果没有基础URL，提示错误
            if not self.base_url:
                self.signal.emit('@An error occurred:API基础URL未设置，请在设置中配置。\n ')
                return
                
            # 确保base_url以斜杠结尾
            if not self.base_url.endswith('/'):
                self.base_url += '/'
                
            # 创建 OpenAI 客户端并请求API
            client = openai.OpenAI(api_key=self.api_key, base_url=self.base_url)
            completion_stream = client.chat.completions.create(
                model=self.model,
                messages=[{"role": "user", "content": self.prompt}],
                stream=True,
            )
            
            response_content = ""
            n = 0
            for chunk in completion_stream:
                # 检查是否被请求中断
                if self.isInterruptionRequested():
                    print("翻译线程收到中断请求，正在停止...")
                    return
                    
                try:
                    content = chunk.choices[0].delta.content or ""
                    response_content += content
                    if n < 3 or n % 3 == 0:
                        self.signal.emit(response_content)
                    n += 1
                except Exception as e:
                    print(f"Error processing chunk: {e}")
                    print(f"Chunk content: {chunk}")
                    
            print(f"Translated text\n----{response_content}\n----")
            self.signal.emit(response_content)
            # 保存到缓存
            self.cache.set(self.text, response_content)
            
        except Exception as e:
            # 只有在非主动中断的情况下才报错
            if not self.isInterruptionRequested():
                print(f"Translation error: {e}")
                self.signal.emit(f'@An error occurred:{str(e)}\n ')

File: src-tauri/src/services/translator.rs
use futures::StreamExt;
use serde_json::Value;
use tauri::Emitter;

pub fn format_prompt(
    template: &str,
    source_language: &str,
    target_language: &str,
    text: &str,
) -> String {
    template
        .replace("{source_language}", source_language)
        .replace("{target_language}", target_language)
        .replace("{selected_text}", text)
}

/// Parse one SSE data line, returning the content delta if present.
pub fn parse_sse_line(line: &str) -> Option<String> {
    let data = line.strip_prefix("data: ")?;
    if data == "[DONE]" {
        return None;
    }
    let json: Value = serde_json::from_str(data).ok()?;
    json["choices"][0]["delta"]["content"]
        .as_str()
        .map(|s| s.to_string())
}

/// Stream a translation from an OpenAI-compatible API, emitting chunks via Tauri events.
/// Returns the full accumulated translation on success.
pub async fn translate_stream(
    text: &str,
    prompt_template: &str,
    source_language: &str,
    target_language: &str,
    api_key: &str,
    base_url: &str,
    model: &str,
    app_handle: &tauri::AppHandle,
) -> Result<String, String> {
    let prompt = format_prompt(prompt_template, source_language, target_language, text);

    let body = serde_json::json!({
        "model": model,
        "messages": [
            { "role": "user", "content": prompt }
        ],
        "stream": true
    });

    let url = format!(
        "{}/chat/completions",
        base_url.trim_end_matches('/')
    );

    let client = reqwest::Client::builder()
        .connect_timeout(std::time::Duration::from_secs(10))
        .timeout(std::time::Duration::from_secs(60))
        .build()
        .map_err(|e| format!("Failed to create HTTP client: {}", e))?;
    let response = client
        .post(&url)
        .header("Authorization", format!("Bearer {}", api_key))
        .header("Content-Type", "application/json")
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("Network error: {}", e))?;

    if !response.status().is_success() {
        let status = response.status();
        let body_text = response.text().await.unwrap_or_default();
        return Err(format!("API error ({}): {}", status, body_text));
    }

    let mut stream = response.bytes_stream();
    let mut accumulated = String::new();
    let mut buffer = String::new();

    while let Some(chunk) = stream.next().await {
        let chunk = chunk.map_err(|e| format!("Stream error: {}", e))?;
        buffer.push_str(&String::from_utf8_lossy(&chunk));

        while let Some(line_end) = buffer.find('\n') {
            let line = buffer[..line_end].trim().to_string();
            buffer = buffer[line_end + 1..].to_string();

            if line.is_empty() {
                continue;
            }

            if let Some(content) = parse_sse_line(&line) {
                accumulated.push_str(&content);
                let _ = app_handle.emit("translation:chunk", &accumulated);
            }
        }
    }

    let _ = app_handle.emit("translation:complete", &accumulated);
    Ok(accumulated)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_format_prompt() {
        let result = format_prompt(
            "Translate from {source_language} to {target_language}: {selected_text}",
            "英语",
            "中文",
            "hello world",
        );
        assert_eq!(result, "Translate from 英语 to 中文: hello world");
    }

    #[test]
    fn test_format_prompt_default_template() {
        let result = format_prompt(
            crate::types::DEFAULT_PROMPT,
            "English",
            "中文",
            "test input",
        );
        assert!(result.contains("English"));
        assert!(result.contains("中文"));
        assert!(result.contains("test input"));
    }

    #[test]
    fn test_parse_sse_line_content() {
        let line = r#"data: {"choices":[{"delta":{"content":"Hello"}}]}"#;
        assert_eq!(parse_sse_line(line), Some("Hello".into()));
    }

    #[test]
    fn test_parse_sse_line_done() {
        assert_eq!(parse_sse_line("data: [DONE]"), None);
    }

    #[test]
    fn test_parse_sse_line_no_content() {
        let line = r#"data: {"choices":[{"delta":{}}]}"#;
        assert_eq!(parse_sse_line(line), None);
    }

    #[test]
    fn test_parse_sse_line_not_data() {
        assert_eq!(parse_sse_line("event: ping"), None);
    }
}
